docs/permission/plugin-authors/05-frontend-authorization.md

```diff
@@ -0,0 +1,167 @@
+    const title = useRef('');
+    const [key, refetchTodos] = useReducer(i => i + 1, 0);
+    const [editElement, setEdit] = useState<Todo | undefined>();
+
++   const { loading: loadingPermission, allowed: canAddTodo } = usePermission({ permission: todoListCreate });
+...
+
+            <Box
+              component="span"
+              alignItems="flex-end"
+              display="flex"
+              flexDirection="row"
+            >
+              <TextField
+                placeholder="Write something here..."
+                onChange={e => (title.current = e.target.value)}
+              />
+-             <Button variant="contained" onClick={handleAdd}>
+-               Add
+-             </Button>
++             {!loadingPermission && (
++               <Button disabled={!canAddTodo} variant="contained" onClick={handleAdd}>
++                 Add
++               </Button>
++             )}
+            </Box>
+...
+```
+
+Here we are using the [`usePermission` hook](https://backstage.io/docs/reference/plugin-permission-react.usepermission) to communicate with the permission policy and receive a decision on whether this user is authorized to create a todo list item.
+
+It's really that simple! Let's change our policy to test the disabled button:
+
+```diff
+// packages/backend/src/plugins/permission.ts
+
+...
+
+    if (isPermission(request.permission, todoListCreate)) {
+      return {
+-       result: AuthorizeResult.ALLOW,
++       result: AuthorizeResult.DENY,
+      };
+    }
+
+...
+```
+
+And now you should see that you are not able to create a todo item from the frontend!
+
+## Using `RequirePermission`
+
+Providing a disabled state can be a helpful signal to users, but there may be cases where hiding the element is preferred. For such cases, you can use the provided [`RequirePermission` component](https://backstage.io/docs/reference/plugin-permission-react.requirepermission):
+
+```diff
+// plugins/todo-list/src/components/TodoListPage/TodoListPage.tsx
+
+...
+  import {
+    alertApiRef,
+    discoveryApiRef,
+    fetchApiRef,
+    useApi,
+  } from '@backstage/core-plugin-api';
+- import { usePermission } from '@backstage/plugin-permission-react';
++ import { RequirePermission } from '@backstage/plugin-permission-react';
+  import { todoListCreate } from '@internal/plugin-todo-list-common';
+...
+
+  export const TodoListPage = () => {
+    const discoveryApi = useApi(discoveryApiRef);
+    const { fetch } = useApi(fetchApiRef);
+    const alertApi = useApi(alertApiRef);
+    const title = useRef('');
+    const [key, refetchTodos] = useReducer(i => i + 1, 0);
+    const [editElement, setEdit] = useState<Todo | undefined>();
+
+-   const { loading: loadingPermission, allowed: canAddTodo } = usePermission({ permission: todoListCreate });
+...
+
+        <Grid container spacing={3} direction="column">
+-         <Grid item>
+-           <Typography variant="body1">Add todo</Typography>
```

Comment:
could have saved yourselves a fair bit of diff churn if you'd had a separate `<TodoPageContents />` component that was all you rendered inside, while you were showing iterations of the surrounding stuff

Reply:
Yep +1, this hurts the readability quite a bit. Always tricky to see what actually changed in a huge diff like this

Reply:
Good idea! Seems like a change worth making considering how many diffs we have throughout the tutorial

Reply:
Done!